aten/src/ATen/native/cuda/Activation.cu

```diff
@@ -243,6 +247,170 @@ std::tuple<Tensor, Tensor> prelu_backward_cuda(const Tensor& grad_out_, const Te
+    auto rand = random_func(&state);
+    #pragma unroll
+    for (int ii = 0; ii < unroll_factor; ii++) {
+      int li = linear_index + blockDim.x * gridDim.x * ii;
+      if (li >= numel) {
+        continue;
+      }
+      scalar_t r = static_cast<scalar_t>((&rand.x)[ii]);
+      r = r * range + lower;
+      if (input[li] <= 0) {
+        output[li] = input[li] * r;
+        noise[li] = r;
+      } else {
+        output[li] = input[li];
+        noise[li] = static_cast<scalar_t>(0);
+      }
+    }
+    __syncthreads();
+  }
+}
+
+template <typename scalar_t>
+inline void _rrelu_with_noise_cuda_train(
+    Tensor& output,
+    const Tensor& input,
+    const Tensor& noise,
+    const Scalar& lower_,
+    const Scalar& upper_,
+    c10::optional<Generator> generator) {
+
+  int64_t numel = input.numel();
+  auto execution_policy = calc_execution_policy(numel);
+
+  auto counter_offset = std::get<0>(execution_policy);
+  auto grid = std::get<1>(execution_policy);
+  auto block = std::get<2>(execution_policy);
+
+  auto gen = get_generator_or_default<CUDAGeneratorImpl>(
+      generator, cuda::detail::getDefaultCUDAGenerator());
+  PhiloxCudaState rng_engine_inputs;
+  {
+    // See Note [Acquire lock when using random generators]
+    std::lock_guard<std::mutex> lock(gen->mutex_);
+    rng_engine_inputs = gen->philox_cuda_state(counter_offset);
+  }
+
+  scalar_t* input_data = input.data_ptr<scalar_t>();
+  scalar_t* noise_data = noise.data_ptr<scalar_t>();
+  scalar_t* output_data = output.data_ptr<scalar_t>();
+
+  double lower = lower_.to<double>();
+  double upper = upper_.to<double>();
+
+  auto stream = at::cuda::getCurrentCUDAStream();
+
+  if (std::is_same<scalar_t, double>::value) {
+    rrelu_with_noise_cuda_kernel<scalar_t, 2><<<grid, block, 0, stream>>>(
+        numel,
+        rng_engine_inputs,
+        output_data,
+        input_data,
+        noise_data,
+        lower,
+        upper,
+        [] __device__ (curandStatePhilox4_32_10_t* state) {
+          return curand_uniform2_double(state);
+        });
+  } else {
+    // half and float
+    rrelu_with_noise_cuda_kernel<scalar_t, 4><<<grid, block, 0, stream>>>(
+        numel,
+        rng_engine_inputs,
+        output_data,
+        input_data,
+        noise_data,
+        lower, upper,
+        [] __device__ (curandStatePhilox4_32_10_t* state) {
+          return curand_uniform4(state);
+        });
+  }
+  C10_CUDA_KERNEL_LAUNCH_CHECK();
+}
+
+Tensor& rrelu_with_noise_out_cuda(const Tensor& self,
+    const Tensor& noise,
+    const Scalar& lower,
+    const Scalar& upper,
+    bool training,
+    c10::optional<Generator> generator,
+    Tensor& output) {
+  TensorArg self_arg{self, "self", 1}, noise_arg{noise, "noise", 2},
+      output_arg{output, "output", 3};
+  checkAllSameGPU("rrelu_with_noise_out_cuda", {self_arg, noise_arg, output_arg});
+
+  auto input = self.contiguous();
+  auto noise_ = noise.contiguous();
```

Comment:
rrelu_with_noise_out_cuda is a user facing function, which means that `output` can also be discontiguous here.